File: BMI.py
# a class to assign every person from the test data their properties.

class BMI:
    def _init_(self):
        self.fullname = ""  # as string
        self.height = 0.0  # as float
        self.weight = 0.0  # as float
        self.bmi = 0.0  # as float
        self.result = ""  # as string


def open_file(temp_array):
    with open("testdata.txt", "r") as temp_file:
        # opens th file testdata.txt as read only to read the data from it.

        row = 0

        # a variable to loop though each row and column in text file.

        for column in temp_file:
            line_split = column.split(",")

            # sets the rows to be spilt at each comma "," in the text file.

            temp_array[row].fullname = str(line_split[0])
            temp_array[row].height = float(line_split[1])
            temp_array[row].weight = float(line_split[2])

            # sets each property of the array to the data from the text file for the current column.

            row += 1

            # moves to the next row.

    temp_file.close()

    # closes the file.


def result(temp_list):
    for counter in range(len(temp_list)):

        # loops for the length of the list

        if temp_list[counter].bmi < 18.5:
            temp_list[counter].result = "underweight"

        elif 18.5 <= temp_list[counter].bmi < 25:
            temp_list[counter].result = "an ideal weight"

        else:
            temp_list[counter].result = "overweight"

        # sets the property of temp_list.result for each record in the array, depending on the temp_list.bmi of said
        # record.


def write_file(temp_list):
    with open("bmi_results.txt", "w") as temp_file:
        for counter in range(len(temp_list)):
            temp_record = "{}'s height is {} m and their weight is {}kgs and their BMI is {}. " \
                          "Overall they are {}."

            # creates a record to be written

            temp_record = temp_record.format(temp_list[counter].fullname, temp_list[counter].height,
                                             temp_list[counter].weight, temp_list[counter].bmi,
                                             temp_list[counter].result)

            # formats the record in the next line for better clarity.

            temp_file.write(temp_record)
            temp_file.write("\n")

            # writes the record to the text file "bmi_results.txt".

    temp_file.close()


def set_bmi(temp_list):
    for person in range(len(temp_list)):
        height_squared = temp_list[person].height ** 2
        temp_list[person].bmi = temp_list[person].weight / height_squared
        temp_list[person].bmi = round(temp_list[person].bmi, 1)

        # calculates each persons BMI from their height and weight using the equation b = w / h^2. Then this is
        # rounded to 1 decimal places, done in multiple lines for better clarity.


bmi_array = [BMI() for counter in range(5)]

# creates an array with the BMI() class with a length of five

open_file(bmi_array)
set_bmi(bmi_array)
result(bmi_array)
write_file(bmi_array)

File: BMI.cs
using System;
using System.IO;

namespace Calculating_BMI
{
    class Program
    {

        class BMI
        {
            public string fullname;
            public double height;
            public double weight;

            private double bmi;
            private string result;

            // private variables for bmi and result to stop direct access

            public double ReturnBMI()
            {
                return bmi;
            }

            public void SetBMI(double getBMI)
            {
                bmi = getBMI;
            }

            public string ReturnResult()
            {
                return result;
            }

            public void SetResult(string getresult)
            {
                result = getresult;
            }

            // public methods to change and access bmi and result variables.

        }

        static void ParseData(BMI[] array)
        {
            string[] data = File.ReadAllLines("testdata.txt");
            // reads all data in an array at each line.

            for (int x = 0; x < data.Length; x++)
            {
                string[] result = data[x].Split(",");
                // sets the rows to be spilt at each comma "," in the text file.

                array[x] = new BMI();
                array[x].fullname = result[0];
                array[x].height = Convert.ToDouble(result[1]);
                array[x].weight = Convert.ToDouble(result[2]);

                // sets each property of the array to the data from the text file.


                double bmi = Math.Round(array[x].weight / Math.Pow(array[x].height, 2), 1);
                array[x].SetBMI(bmi);
                // calculates and assigns each persons BMI.
            }
        }

        static void Calculate_Result(BMI[] array)
        {

            for (int i = 0; i < array.Length; i++)
            {
                if (array[i].ReturnBMI() < 18)
                {
                    array[i].SetResult("underweight");
                } else if (array[i].ReturnBMI() > 25)
                {
                    array[i].SetResult("overweight");
                } else
                {
                    array[i].SetResult("an ideal weight");
                }
            }

            // uses the BMI to give a result of general weight.
        }

        static void Print(BMI[] array)
        {
            for (int i = 0; i < array.Length; i++)
            {
                Console.WriteLine("Hello {0}! You are {1} with a BMI of {2}.", array[i].fullname, array[i].ReturnResult(), array[i].ReturnBMI());
            }
        }

        static void Write(BMI[] array)
        {
            string[] writeToFile = new string[array.Length];
            // creates an array to write to file.

            for (int i = 0; i < array.Length; i++)
            {
                string text = "Hello ";
                text += Convert.ToString(array[i].fullname);
                text += "! You are ";
                text += Convert.ToString(array[i].ReturnResult());
                text += " with a BMI of ";
                text += Convert.ToString(array[i].ReturnBMI()); 
                text += ". \n";
                writeToFile[i] = text;
                // formats the text into lines

            }
            File.WriteAllLines("bmi_results_csharp.txt", writeToFile);
            // write each line of the array to the file
        }

        static void Main(string[] args)
        {
            var People = new BMI[5];
            // creates a new array with BMI class with a length of five.

            ParseData(People);
            Calculate_Result(People);
            Print(People);
            Write(People);
                  
        }
    }
}